TechnicalDocuments/SnapshotTesting.md

```diff
@@ -0,0 +1,50 @@
+* it stores all references images in a single folder for all tests
+* you have to manually find the failure images and diffs to see what went wrong (or use an app like FBSnapshotsViewer)
+* You always have to explicityly set `recordMode` to `true` even if the test has no reference images yet.
+
+### Pros of `SnapshotTesting`
+* `SnapshotTesting` is a brand new library written in Swift by the [Pointfree](pointfree.co) duo. It's a great example of a protocol based API using struct based protocol witnesses as explained in the Pointfree videos on the subject.
+* It supports snapshotting anything, not just `UIView` or `CALayer`. For example, you could snapshot a `URLRequest` instead of having to laboriously type out the expected values by hand.
+You can snapshot anything as a text representation, for example, a view hierarchy, a Bento render tree, a struct, or a JSON dictionary.
+* It is easy to extend the library to add explicit support for snapshotting of other types, instead of just using the `Any` strategy.
+* Snapshots are stored in a `__Snapshots__` folder in the same folder hierarchy as the test source. This makes it easy to find relevant snapshots.
+* It uses `XCTAttachment`s to store failure snapshots and diffs, which makes these available directly from the Xcode test report navigator, without requiring any other software or needing to hunt for them.
+* It is fully Linux compatiable. Allowing snapshot testing to be used on projects that are not iOS or macOS (eg our merge bot).
+* Even if `record` is set to `false` it will still record reference snapshots if none currently exist, so you only have to set it to `true` when updating snapshots becuase of changes.
+
+### Cons of `SnapshotTesting`
+* It's a very new library and in active development. The developers themselves state *"This library should be considered alpha, and not stable. Breaking changes will happen often."* However, provided we stick to a specific version or commit in the Podfile, this should not be a big issue.
+
+### Benchmarks
+I used `BabylonChatBotUI` for running some rudimentary benchmarks between the two libraries as it seems to have the most snapshot tests and, more importantly, it seems to have *only* snapshot test so nothing else would pollute the times.
+
+Average times for running `Cmd+U` on my machine in Xcode for `BabylonChatBotUI`:
+
+iOSSnapshotTestCase | SnapshotTesting
+-------------------|-----------------
+11 seconds | 15 seconds
+
+So, it would seem that `SnapshotTesting` is roughtly 36% slower than `iOSSnapshotTestCase` 😞 .
+
+I have no idea where the speed difference is. Could well be just that unoptimized Swift is slower than Objective-C 🤷🏼‍♀️ . Still, I don't think this is a reason to reject the benefits of `SnapshotTesting`.
```

Comment:
Thanks for these benchmarks!

I think this could be explained by the fact that we support async snapshots (for snapshotting things that require callbacks or delegates, like `WKWebView`), and so every assertion is using an `XCTWaiter` behind the scenes. We could probably skip that for assertions that we know are not async.

We'll investigate this soon.

Reply:
This will remove some headache from our side as well as we had to implement exactly this case on our side (waiting for web view to load in snapshot tests) 🎉

Reply:
Oh wow, I totally missed the async snapshot support. This might explain why in the few async snapshot scenarios we have using `SnapshotTesting` are causing a crash (`EXC_BAD_ACCESS`). I didn't mention this before as I was trying to nail down the cause a bit better and get an easily reproducible case.

Curiously, the test runs fine, but it's when the XCTest UIApplication is torn down that there is an `objc_msgSend` crash. I will see if switching to the Async API solves this 👍